hera_pspec/uvpspec.py

```diff
@@ -1616,6 +1630,142 @@ def set_cosmology(self, new_cosmo, overwrite=False, new_beam=None,
             self.norm_units = "h^-3 Mpc^3"
 
+    def get_exact_window_functions(self, ftbeam_file='', this_spw=None,
+                                   verbose=False, inplace=True, add_to_history='',
+                                   x_orientation=None):
+        """
+        Obtain the exact window functions corresponding to UVPSpec object.
+
+        Depending on the format of the UVPSpec object given as input, the shape
+        of the output array will be different: (Nblpts, Ndlys, Nkperp, Nkpara, Npols).
+        Weights can be applied to the different baseline pairs to facilitate
+        spherical average later.
+
+        Parameters
+        ----------
+
+        ftbeam_file : str, optional
+            Definition of the beam Fourier transform to be used.
+            Options include;
+                - Root name of the file to use, without the polarisation
+                Ex : FT_beam_HERA_dipole (+ path)
+                - '' for computation from beam simulations (slow)
+
+        this_spw : int, optional
+            Spectral window index. If None, the window functions are computed on 
+            all the uvp.spw_ranges, successively. Default: None.
+
+        verbose : bool, optional
+            If True, print progress, warnings and debugging info to stdout.
+
+        inplace : bool, optional
+            If True (default value), the UVPspec attribute window_function_array is filled with the
+            values computed in the function, and window_function_kperp_bins and
+            window_function_kpara_bins array are added as attributed.
+            It False, returns kperp_bins, kpara_bins and window functions computed.
+            Automatically set to False if blpair_groups is not None (that is,
+            if the window functions are not computed on all the blpairs).
+
+        add_to_history : str, optional
+            Added text to add to file history.
+
+        x_orientation: str, optional
+            Orientation in cardinal direction east or north of X dipole.
+            Default keeps polarization in X and Y basis.
+            Used to convert polstr to polnum and conversely. Default: None.
+        """
+
+        # sets attribute exact_windows to False if not defined
+        # (UVPspec object created with older versions of hera_pspec)
+        try:
+            self.exact_windows
+        except AttributeError:
+            self.exact_windows = False
+        if self.exact_windows and inplace:
+            warnings.warn("Exact window functions already computed, overwriting...")
+
+        blpair_groups, blpair_lens, _ = self.get_red_blpairs()
+
+        # check spw input and create array of spws to loop over
+        if this_spw is None:
+            # if no spw specified, use attribute
+            spws = np.arange(self.Nspws)
+        else:
+            # check if spw given is in uvp
+            assert this_spw in self.spw_array, "input spw is not in UVPSpec.spw_array."
```

Comment:
Shouldn't you allow to set multiple spws since you obviously handle the case of multiple spws?

Reply:
Done :)